In [ ]:
import pandas as pd
import numpy as np
import torch
from transformers import (
    AutoModelForCausalLM,
    AutoTokenizer,
    TrainingArguments,
    Trainer,
    DataCollatorForLanguageModeling
)
from datasets import Dataset
from sklearn.metrics import accuracy_score, precision_recall_fscore_support, roc_auc_score
from sklearn.model_selection import train_test_split
from peft import LoraConfig, get_peft_model, prepare_model_for_kbit_training
import time
import json
import random
from tqdm.auto import tqdm


In [ ]:
# Доли пропущенных признаков для экспериментов
MISSING_RATES = [0.0, 0.2, 0.5, 0.9]

# Настройка промпта для Diabetes
prompt_config = {
        "task": "Predict whether a patient has diabetes",
        "pos_label": "yes",
        "neg_label": "no",
        "entity": "Patient",
        "question": "Does this patient have diabetes?"
    }

base_output_dir = "/content/drive/MyDrive/finetuned_qwen_missing_Diabetes"

# 1. Загрузка данных

In [ ]:
import kagglehub
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder

def load_dataset():
    path = kagglehub.dataset_download("uciml/pima-indians-diabetes-database") + "/diabetes.csv"
    df = pd.read_csv(path)
    target_name = 'Outcome'
    X = df.drop(columns=[target_name])
    y = df[target_name]

    feature_names = X.columns.to_list()

    le = LabelEncoder()
    df[target_name] = le.fit_transform(df[target_name])

    return df, feature_names, target_name


def split_dataset(df, target_name, test_size=0.2, val_size=0.25, seed=42):
    # Разделение на train/val/test (60/20/20)
    train_val_df, test_df = train_test_split(
        df,
        test_size=test_size,
        random_state=seed,
        stratify=df[target_name]
    )

    train_df, val_df = train_test_split(
        train_val_df,
        test_size=val_size,
        random_state=seed,
        stratify=train_val_df[target_name]
    )

    return train_df, val_df, test_df

df, feature_names, target_name = load_dataset()
train_df, val_df, test_df = split_dataset(df, target_name)

100%|██████████| 8.91k/8.91k [00:00<00:00, 13.5MB/s]

Extracting files...


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [ ]:
for name, df in [("train", train_df), ("val", val_df), ("test", test_df)]:
    counts = df[target_name].value_counts()
    pcts   = df[target_name].value_counts(normalize=True) * 100
    print(f"\n{name} (всего: {len(df)}):")
    print(f"  {prompt_config['neg_label']}: {counts[0]:5d} — {pcts[0]:.1f}%")
    print(f"  {prompt_config['pos_label']}: {counts[1]:5d} — {pcts[1]:.1f}%")


train (всего: 460):
  no:   300 — 65.2%
  yes:   160 — 34.8%

val (всего: 154):
  no:   100 — 64.9%
  yes:    54 — 35.1%

test (всего: 154):
  no:   100 — 64.9%
  yes:    54 — 35.1%


# 2. Вспомогательные функции

In [ ]:
def row_to_text_template_with_missing(row, feature_names, target_name, missing_rate=0.0, seed=None):

    rng = np.random.default_rng(seed)

    # Определяем, какие признаки пропустить
    n_drop = int(len(feature_names) * missing_rate)
    dropped = set(rng.choice(feature_names, size=n_drop, replace=False)) if n_drop > 0 else set()

    template_parts = []

    for feature in feature_names:
        if feature in dropped:
            continue  # пропускаем признак

        value = row[feature]

        if isinstance(value, (int, np.integer)):
            phrase = f"The value of {feature} is {value}."
        elif isinstance(value, (float, np.floating)):
            phrase = f"The value of {feature} is {value:.2f}."
        else:
            phrase = f"The category of {feature} is {value}."

        template_parts.append(phrase)

    text = " ".join(template_parts)

    return text

# Тест
test_row = train_df.iloc[0]
display(train_df.head(1))
for rate in MISSING_RATES:
    text_output = row_to_text_template_with_missing(
        row=test_row,
        feature_names=feature_names,
        target_name=target_name,
        missing_rate=rate
    )

    print(f"Missing Rate: {rate * 100:.0f}%")
    print(text_output)

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
550,1,116,70,28,0,27.4,0.204,21,0


Missing Rate: 0%
The value of Pregnancies is 1.00. The value of Glucose is 116.00. The value of BloodPressure is 70.00. The value of SkinThickness is 28.00. The value of Insulin is 0.00. The value of BMI is 27.40. The value of DiabetesPedigreeFunction is 0.20. The value of Age is 21.00.
Missing Rate: 20%
The value of Pregnancies is 1.00. The value of Glucose is 116.00. The value of BloodPressure is 70.00. The value of SkinThickness is 28.00. The value of Insulin is 0.00. The value of BMI is 27.40. The value of Age is 21.00.
Missing Rate: 50%
The value of Glucose is 116.00. The value of SkinThickness is 28.00. The value of BMI is 27.40. The value of Age is 21.00.
Missing Rate: 90%
The value of Age is 21.00.


In [ ]:
def parse_prediction(response, prompt_config):
    response = response.lower().strip().rstrip('.,!? ')
    pos, neg = prompt_config['pos_label'].lower(), prompt_config['neg_label'].lower()

    # Точное совпадение
    if response == pos: return 1
    if response == neg: return 0

    # Начинается с метки
    if response.startswith(pos): return 1
    if response.startswith(neg): return 0

    # Содержит метку как отдельное слово
    words = response.split()
    if pos in words: return 1
    if neg in words: return 0

    # Не удалось распознать
    print(f"Warning: Could not parse '{response}' (expected '{prompt_config['pos_label']}' or '{prompt_config['neg_label']}')")
    return 0



response = prompt_config['pos_label']
pred = parse_prediction(response, prompt_config)
print(f"Response: '{response}'\nPrediction: {pred}\n")

response = "hi"
pred = parse_prediction(response, prompt_config)
print(f"Response: '{response}'\nPrediction: {pred}")

Response: 'yes'
Prediction: 1

Response: 'hi'
Prediction: 0


In [ ]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score
# Вычисление метрик качества
def compute_metrics(y_true, y_pred, y_prob=None):
    roc = roc_auc_score(y_true, y_prob if y_prob is not None else y_pred)
    return {
        "ROC-AUC": roc_auc_score(y_true, y_prob if y_prob is not None else y_pred),
        "F1":       f1_score(y_true, y_pred, zero_division=0),
        "Accuracy": accuracy_score(y_true, y_pred),
        "Precision": precision_score(y_true, y_pred, zero_division=0),
        "Recall":   recall_score(y_true, y_pred, zero_division=0),
    }

In [ ]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score
from sklearn.utils import resample

def bootstrap_metrics(y_true, y_pred, y_prob=None, n_iter=1000):
    scores = []

    for i in range(n_iter):
        # Bootstrap выборка
        if y_prob is not None:
            y_true_boot, y_pred_boot, y_prob_boot = resample(
                y_true, y_pred, y_prob, random_state=i+1
            )
        else:
            y_true_boot, y_pred_boot = resample(
                y_true, y_pred, random_state=i+1
            )
            y_prob_boot = None

        try:
            # Вычисление метрик
            if y_prob_boot is not None:
                auc = roc_auc_score(y_true_boot, y_prob_boot)
            else:
                auc = 0.0

            f1 = f1_score(y_true_boot, y_pred_boot, zero_division=0)
            pr = precision_score(y_true_boot, y_pred_boot, zero_division=0)
            rc = recall_score(y_true_boot, y_pred_boot, zero_division=0)

            acc = accuracy_score(y_true_boot, y_pred_boot)
            scores.append((auc, f1, acc, pr, rc))

        except ValueError:
            # Пропускание сэмплов, где не представлены все классы
            continue

    scores = np.asarray(scores)
    means, stds = scores.mean(0), scores.std(0, ddof=1)
    names = ["ROC-AUC", "F1", "Accuracy", "Precision", "Recall"]

    return {n: f"{m:.4f}±{s:.4f}" for n, m, s in zip(names, means, stds)}

# Тестирование функции
sample_row = test_df.iloc[0]
sample_text = row_to_text_template_with_missing(sample_row, feature_names, target_name)
print(f"\nПример текста:\n{sample_text[:300]}")


Пример текста:
The value of Pregnancies is 7.00. The value of Glucose is 159.00. The value of BloodPressure is 64.00. The value of SkinThickness is 0.00. The value of Insulin is 0.00. The value of BMI is 27.40. The value of DiabetesPedigreeFunction is 0.29. The value of Age is 40.00.


In [ ]:
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer, BitsAndBytesConfig

#Загрузка модели Qwen 3.0 4B Instruct
model_name = "Qwen/Qwen3-4B-Instruct-2507"

device = "cuda" if torch.cuda.is_available() else "cpu"
print(f"Using device: {device}")

# Загрузка токенизаторов
tokenizer = AutoTokenizer.from_pretrained(model_name)

# Загрузка модель с 16-битной квантизацией и автоматическим распределением по устройствам
base_model = AutoModelForCausalLM.from_pretrained(
    model_name,
    dtype=torch.bfloat16,
    device_map=device, # Explicitly force all model layers to the detected device (GPU if available)
    attn_implementation="sdpa"
)

if torch.cuda.is_available():
    print(f"Использовано памяти: {torch.cuda.memory_allocated() / 1e9:.2f} GB")

Using device: cuda


/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/727 [00:00<?, ?B/s]

tokenizer_config.json: 0.00B [00:00, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json:   0%|          | 0.00/11.4M [00:00<?, ?B/s]

model.safetensors.index.json: 0.00B [00:00, ?B/s]

Fetching 3 files:   0%|          | 0/3 [00:00<?, ?it/s]

Loading weights:   0%|          | 0/398 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/238 [00:00<?, ?B/s]

Использовано памяти: 8.04 GB


In [ ]:
def create_prompt_missing(row, feature_names, target_name, prompt_config, tokenizer, missing_rate=0.0, seed=None):

    system_prompt = (
        f"You are a classifier. {prompt_config['task']}: "
        f"'{prompt_config['pos_label']}' or '{prompt_config['neg_label']}'. "
        f"Answer with only one word: '{prompt_config['pos_label']}' or '{prompt_config['neg_label']}'."
    )

    input_text = row_to_text_template_with_missing(row, feature_names, target_name, missing_rate, seed)

    messages = [
        {"role": "system", "content": system_prompt},
        {"role": "user",
         "content": f"{prompt_config['entity']} information: {input_text}\n{prompt_config['question']}"},
    ]

    return tokenizer.apply_chat_template(
        messages,
        tokenize=False,
        add_generation_prompt=True # добавление role assistant
    )

In [ ]:
import torch.nn.functional as F

def predict_single_with_prob(prompt, prompt_config, model, tokenizer, device, max_new_tokens=5):

    model_inputs = tokenizer([prompt], return_tensors="pt").to(device)

    with torch.no_grad():
        outputs = model.generate(
            model_inputs.input_ids,
            max_new_tokens=max_new_tokens,
            do_sample=False,
            pad_token_id=tokenizer.eos_token_id,
            output_scores=True,
            return_dict_in_generate=True
        )

    # Декодирование текста
    generated_ids = outputs.sequences[0][model_inputs.input_ids.shape[1]:]
    response = tokenizer.decode(generated_ids, skip_special_tokens=True).strip().lower()

    # Извлечение вероятностей для меток из prompt_config
    first_token_logits = outputs.scores[0][0]

    pos_id = tokenizer.encode(prompt_config['pos_label'], add_special_tokens=False)[0]
    neg_id = tokenizer.encode(prompt_config['neg_label'], add_special_tokens=False)[0]

    pos_logit = first_token_logits[pos_id]
    neg_logit = first_token_logits[neg_id]

    probs = F.softmax(torch.stack([pos_logit, neg_logit]), dim=0)
    prob_pos = probs[0].item()

    return response, prob_pos

# Тест
test_prompt = create_prompt_missing(test_df.iloc[0], feature_names, target_name, prompt_config, tokenizer)
response, prob = predict_single_with_prob(test_prompt, prompt_config, base_model, tokenizer, device)
print(f"Response: '{response}', Probability: {prob:.4f}")

The following generation flags are not valid and may be ignored: ['temperature', 'top_p', 'top_k']. Set `TRANSFORMERS_VERBOSITY=info` for more details.
The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.


Response: 'yes', Probability: 0.9047


# 3. Fine-tuning с LoRA

## Подготовка данных для Fine-tuning

In [ ]:
def balance_dataset(train_df, target_name, prompt_config, seed=42):
    df_0 = train_df[train_df[target_name] == 0]
    df_1 = train_df[train_df[target_name] == 1]

    df_majority = df_0 if len(df_0) > len(df_1) else df_1
    df_minority = df_1 if len(df_0) > len(df_1) else df_0

    print(f"До балансировки:")
    print(f"{prompt_config['neg_label']}:  {len(df_majority)}")
    print(f"{prompt_config['pos_label']}: {len(df_minority)}")

    df_minority_up = resample(df_minority, replace=True,
                              n_samples=len(df_majority), random_state=seed)

    train_df_balanced = pd.concat([df_majority, df_minority_up])
    train_df_balanced = train_df_balanced.sample(frac=1, random_state=seed).reset_index(drop=True)

    print(f"\nПосле балансировки:")
    print(train_df_balanced[target_name].value_counts())

    return train_df_balanced

train_df_balanced = balance_dataset(train_df, target_name, prompt_config)

До балансировки:
no:  300
yes: 160

После балансировки:
Outcome
0    300
1    300
Name: count, dtype: int64


In [ ]:
# Создание текстов для обучения

def create_training_dataset(df, feature_names, target_name, prompt_config, tokenizer, missing_rate=0.0):
    system_prompt = (
        f"You are a classifier. {prompt_config['task']}: "
        f"'{prompt_config['pos_label']}' or '{prompt_config['neg_label']}'. "
        f"Answer with only one word: '{prompt_config['pos_label']}' or '{prompt_config['neg_label']}'."
    )
    texts = []

    for idx, (_, row) in enumerate(tqdm(df.iterrows(), total=len(df), desc=f"Dataset (missing={missing_rate:.0%})")):
        input_text = row_to_text_template_with_missing(row, feature_names, target_name, missing_rate, idx)
        target = prompt_config['pos_label'] if row[target_name] == 1 else prompt_config['neg_label']

        messages = [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": f"{prompt_config['entity']} information: {input_text}\n{prompt_config['question']}"},
            {"role": "assistant", "content": target}
        ]

        text = tokenizer.apply_chat_template(
            messages,
            tokenize=False,
            add_generation_prompt=False
        )
        texts.append(text)

    return texts


In [ ]:
def finetune_model(missing_rate, train_df_balanced, feature_names, target_name,
                   prompt_config, tokenizer, base_model, device,
                   num_epochs=3, batch_size=16):
    """
    Полный цикл fine-tuning для заданного missing_rate.
    Возвращает путь к лучшему checkpoint.
    """
    output_dir = f"{base_output_dir}_missing{int(missing_rate * 100):03d}"
    print(f"  Fine-tuning: missing_rate = {missing_rate:.0%}")
    print(f"  Output dir : {output_dir}")

    # Датасет
    train_texts = create_training_dataset(
        train_df_balanced, feature_names, target_name,
        prompt_config, tokenizer, missing_rate=missing_rate)
    train_dataset = Dataset.from_dict({"text": train_texts})

    def tokenize_fn(examples):
        return tokenizer(examples["text"], truncation=True, max_length=512, padding=False)

    tokenized_dataset = train_dataset.map(tokenize_fn, batched=True,
                                          remove_columns=train_dataset.column_names)

    # LoRA
    lora_config = LoraConfig(
        r=16, lora_alpha=32,
        target_modules=["q_proj", "k_proj", "v_proj", "o_proj",
                        "gate_proj", "up_proj", "down_proj"],
        lora_dropout=0.05, bias="none", task_type="CAUSAL_LM"
    )
    model_lora = get_peft_model(base_model, lora_config)
    model_lora.print_trainable_parameters()

    training_args = TrainingArguments(
        output_dir=output_dir,
        num_train_epochs=num_epochs,
        per_device_train_batch_size=batch_size,
        gradient_accumulation_steps=2,
        learning_rate=2e-4,
        bf16=True,
        logging_steps=10,
        save_strategy="epoch",
        optim="adamw_torch_fused",
        warmup_steps=50,
        max_grad_norm=1.0,
        weight_decay=0.01,
        report_to="none",
        dataloader_num_workers=4,
        group_by_length=True,
    )

    trainer = Trainer(
        model=model_lora,
        args=training_args,
        train_dataset=tokenized_dataset,
        data_collator=DataCollatorForLanguageModeling(tokenizer=tokenizer, mlm=False)
    )

    t0 = time.time()
    trainer.train()
    print(f"Обучение завершено за {(time.time()-t0)/60:.1f} мин")

    trainer.save_model(output_dir)

    # Возвращаем базовую модель без LoRA весов (важно для следующего эксперимента)
    model_lora = model_lora.unload()
    torch.cuda.empty_cache()

    return output_dir

In [ ]:
import glob
import re
from peft import PeftModel

def evaluate_checkpoints_on_val(output_dir, val_df, feature_names, target_name,
                                 prompt_config, tokenizer, base_model, device,
                                 eval_missing_rate=0.0):

    checkpoints = sorted(
        glob.glob(f"{output_dir}/checkpoint-*"),
        key=lambda x: int(re.search(r'checkpoint-(\d+)', x).group(1))
    )
    print(f"Найдено {len(checkpoints)} checkpoints в {output_dir}")

    best_auc, best_checkpoint = 0.0, None
    for cp in checkpoints:
        model_eval = PeftModel.from_pretrained(base_model, cp).eval()

        y_true, y_prob = [], []
        for _, row in tqdm(val_df.iterrows(), total=len(val_df),
                           desc=f"Val eval {cp.split('/')[-1]}"):
            prompt = create_prompt_missing(
                row, feature_names, target_name, prompt_config, tokenizer,
                missing_rate=eval_missing_rate)
            _, prob = predict_single_with_prob(prompt, prompt_config, model_eval, tokenizer, device)
            y_true.append(row[target_name])
            y_prob.append(prob)

        y_pred = [1 if p > 0.5 else 0 for p in y_prob]
        auc = roc_auc_score(y_true, y_prob)
        print(f"  {cp.split('/')[-1]}  ROC-AUC={auc:.4f}")

        if auc > best_auc:
            best_auc = auc
            best_checkpoint = cp

        del model_eval
        torch.cuda.empty_cache()

    print(f"Лучший checkpoint: {best_checkpoint}  (ROC-AUC={best_auc:.4f})")
    return best_checkpoint

In [ ]:
def evaluate_on_test(best_checkpoint, test_df, feature_names, target_name,
                     prompt_config, tokenizer, base_model, device,
                     eval_missing_rate=0.0):
    model_finetuned = PeftModel.from_pretrained(base_model, best_checkpoint).eval()

    y_true, y_pred, y_prob = [], [], []
    t0 = time.time()

    for _, row in tqdm(test_df.iterrows(), total=len(test_df), desc="Test eval"):
        prompt = create_prompt_missing(
            row, feature_names, target_name, prompt_config, tokenizer,
            missing_rate=eval_missing_rate)
        response, prob = predict_single_with_prob(
            prompt, prompt_config, model_finetuned, tokenizer, device)
        y_true.append(row[target_name])
        y_pred.append(parse_prediction(response, prompt_config))
        y_prob.append(prob)

    elapsed = time.time() - t0
    metrics = compute_metrics(np.array(y_true), np.array(y_pred), np.array(y_prob))
    boot = bootstrap_metrics(np.array(y_true), np.array(y_pred), np.array(y_prob), n_iter=1000)

    print("Метрики:")
    for k, v in metrics.items():
        print(f"  {k}: {v:.4f}")
    print("Bootstrap (mean±std):")
    for k, v in boot.items():
        print(f"  {k}: {v}")

    del model_finetuned
    torch.cuda.empty_cache()

    return {
        "metrics": metrics,
        "bootstrap": boot,
        "time_total": elapsed,
        "time_per_sample": elapsed / len(y_true),
        "best_checkpoint": best_checkpoint,
        "eval_missing_rate": eval_missing_rate,
    }

In [ ]:
all_results = {}

start_time = time.time()

for missing_rate in MISSING_RATES:
    tag = f"missing_{int(missing_rate * 100):03d}pct"
    print(f"Эксперимент: missing_rate = {missing_rate:.0%}")

    # 7.1 Fine-tuning
    output_dir = finetune_model(
        missing_rate=missing_rate,
        train_df_balanced=train_df_balanced,
        feature_names=feature_names,
        target_name=target_name,
        prompt_config=prompt_config,
        tokenizer=tokenizer,
        base_model=base_model,
        device=device,
        num_epochs=20,
        batch_size=16,
    )

    # 7.2 Выбор лучшего checkpoint по val
    best_cp = evaluate_checkpoints_on_val(
        output_dir=output_dir,
        val_df=val_df,
        feature_names=feature_names,
        target_name=target_name,
        prompt_config=prompt_config,
        tokenizer=tokenizer,
        base_model=base_model,
        device=device,
        eval_missing_rate=missing_rate,
    )

    # 7.3 Оценка на test
    result = evaluate_on_test(
        best_checkpoint=best_cp,
        test_df=test_df,
        feature_names=feature_names,
        target_name=target_name,
        prompt_config=prompt_config,
        tokenizer=tokenizer,
        base_model=base_model,
        device=device,
        eval_missing_rate=missing_rate,
    )

    all_results[tag] = result

print(f"Эксперимент завершен за {(time.time()-start_time)/60:.1f} мин")

Эксперимент: missing_rate = 0%
  Fine-tuning: missing_rate = 0%
  Output dir : /content/drive/MyDrive/finetuned_qwen_missing_Diabetes_missing000


Dataset (missing=0%):   0%|          | 0/600 [00:00<?, ?it/s]

Map:   0%|          | 0/600 [00:00<?, ? examples/s]

trainable params: 33,030,144 || all params: 4,055,498,240 || trainable%: 0.8145


Step,Training Loss
10,1.742849
20,0.858266
30,0.263810
40,0.204728
50,0.192795
60,0.195388
70,0.190291
80,0.188064
90,0.189811
100,0.181685


Обучение завершено за 7.2 мин
Найдено 20 checkpoints в /content/drive/MyDrive/finetuned_qwen_missing_Diabetes_missing000


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-19:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-19  ROC-AUC=0.8102


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-38:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-38  ROC-AUC=0.8580


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-57:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-57  ROC-AUC=0.8442


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-76:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-76  ROC-AUC=0.8503


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-95:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-95  ROC-AUC=0.8551


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-114:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-114  ROC-AUC=0.8581


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-133:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-133  ROC-AUC=0.8592


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-152:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-152  ROC-AUC=0.8449


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-171:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-171  ROC-AUC=0.8463


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-190:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-190  ROC-AUC=0.8396


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-209:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-209  ROC-AUC=0.8287


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-228:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-228  ROC-AUC=0.8279


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-247:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-247  ROC-AUC=0.8116


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-266:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-266  ROC-AUC=0.8389


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-285:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-285  ROC-AUC=0.8395


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-304:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-304  ROC-AUC=0.8277


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-323:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-323  ROC-AUC=0.8190


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-342:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-342  ROC-AUC=0.8252


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-361:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-361  ROC-AUC=0.8204


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-380:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-380  ROC-AUC=0.8214
Лучший checkpoint: /content/drive/MyDrive/finetuned_qwen_missing_Diabetes_missing000/checkpoint-133  (ROC-AUC=0.8592)


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Test eval:   0%|          | 0/154 [00:00<?, ?it/s]

Метрики:
  ROC-AUC: 0.8039
  F1: 0.6333
  Accuracy: 0.7143
  Precision: 0.5758
  Recall: 0.7037
Bootstrap (mean±std):
  ROC-AUC: 0.8037±0.0346
  F1: 0.6311±0.0524
  Accuracy: 0.7143±0.0364
  Precision: 0.5761±0.0621
  Recall: 0.7026±0.0621
Эксперимент: missing_rate = 20%
  Fine-tuning: missing_rate = 20%
  Output dir : /content/drive/MyDrive/finetuned_qwen_missing_Diabetes_missing020


Dataset (missing=20%):   0%|          | 0/600 [00:00<?, ?it/s]

Map:   0%|          | 0/600 [00:00<?, ? examples/s]

/usr/local/lib/python3.12/dist-packages/peft/mapping_func.py:72: UserWarning: You are trying to modify a model with PEFT for a second time. If you want to reload the model with a different config, make sure to call `.unload()` before.
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


trainable params: 33,030,144 || all params: 4,055,498,240 || trainable%: 0.8145


Step,Training Loss
10,1.934169
20,0.955768
30,0.295889
40,0.215734
50,0.202916
60,0.205032
70,0.201147
80,0.197927
90,0.200594
100,0.191751


Обучение завершено за 6.7 мин
Найдено 20 checkpoints в /content/drive/MyDrive/finetuned_qwen_missing_Diabetes_missing020


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-19:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-19  ROC-AUC=0.7940


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-38:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-38  ROC-AUC=0.8312


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-57:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-57  ROC-AUC=0.8281


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-76:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-76  ROC-AUC=0.8731


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-95:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-95  ROC-AUC=0.8298


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-114:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-114  ROC-AUC=0.8749


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-133:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-133  ROC-AUC=0.8427


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-152:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-152  ROC-AUC=0.8156


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-171:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-171  ROC-AUC=0.8187


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-190:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-190  ROC-AUC=0.8245


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-209:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-209  ROC-AUC=0.8099


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-228:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-228  ROC-AUC=0.8257


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-247:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-247  ROC-AUC=0.7651


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-266:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-266  ROC-AUC=0.8095


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-285:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-285  ROC-AUC=0.7970


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-304:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-304  ROC-AUC=0.8607


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-323:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-323  ROC-AUC=0.8279


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-342:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-342  ROC-AUC=0.8239


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-361:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-361  ROC-AUC=0.7642


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-380:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-380  ROC-AUC=0.7931
Лучший checkpoint: /content/drive/MyDrive/finetuned_qwen_missing_Diabetes_missing020/checkpoint-114  (ROC-AUC=0.8749)


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Test eval:   0%|          | 0/154 [00:00<?, ?it/s]

Метрики:
  ROC-AUC: 0.7559
  F1: 0.6383
  Accuracy: 0.6688
  Precision: 0.5172
  Recall: 0.8333
Bootstrap (mean±std):
  ROC-AUC: 0.7549±0.0381
  F1: 0.6351±0.0487
  Accuracy: 0.6669±0.0386
  Precision: 0.5156±0.0550
  Recall: 0.8324±0.0508
Эксперимент: missing_rate = 50%
  Fine-tuning: missing_rate = 50%
  Output dir : /content/drive/MyDrive/finetuned_qwen_missing_Diabetes_missing050


Dataset (missing=50%):   0%|          | 0/600 [00:00<?, ?it/s]

Map:   0%|          | 0/600 [00:00<?, ? examples/s]

/usr/local/lib/python3.12/dist-packages/peft/mapping_func.py:72: UserWarning: You are trying to modify a model with PEFT for a second time. If you want to reload the model with a different config, make sure to call `.unload()` before.
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


trainable params: 33,030,144 || all params: 4,055,498,240 || trainable%: 0.8145


Step,Training Loss
10,2.566033
20,1.216911
30,0.319788
40,0.206747
50,0.195288
60,0.193403
70,0.188242
80,0.187708
90,0.189873
100,0.182376


Обучение завершено за 5.6 мин
Найдено 20 checkpoints в /content/drive/MyDrive/finetuned_qwen_missing_Diabetes_missing050


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-19:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-19  ROC-AUC=0.7251


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-38:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-38  ROC-AUC=0.7563


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-57:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-57  ROC-AUC=0.7235


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-76:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-76  ROC-AUC=0.8073


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-95:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-95  ROC-AUC=0.7740


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-114:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-114  ROC-AUC=0.8323


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-133:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-133  ROC-AUC=0.7791


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-152:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-152  ROC-AUC=0.7399


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-171:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-171  ROC-AUC=0.7457


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-190:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-190  ROC-AUC=0.7929


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-209:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-209  ROC-AUC=0.7161


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-228:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-228  ROC-AUC=0.7793


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-247:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-247  ROC-AUC=0.7564


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-266:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-266  ROC-AUC=0.7121


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-285:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-285  ROC-AUC=0.7069


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-304:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-304  ROC-AUC=0.7060


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-323:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-323  ROC-AUC=0.7219


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-342:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-342  ROC-AUC=0.6956


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-361:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-361  ROC-AUC=0.7343


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-380:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-380  ROC-AUC=0.7138
Лучший checkpoint: /content/drive/MyDrive/finetuned_qwen_missing_Diabetes_missing050/checkpoint-114  (ROC-AUC=0.8323)


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Test eval:   0%|          | 0/154 [00:00<?, ?it/s]

Метрики:
  ROC-AUC: 0.7796
  F1: 0.5862
  Accuracy: 0.6883
  Precision: 0.5484
  Recall: 0.6296
Bootstrap (mean±std):
  ROC-AUC: 0.7799±0.0357
  F1: 0.5846±0.0540
  Accuracy: 0.6886±0.0363
  Precision: 0.5488±0.0630
  Recall: 0.6302±0.0651
Эксперимент: missing_rate = 90%
  Fine-tuning: missing_rate = 90%
  Output dir : /content/drive/MyDrive/finetuned_qwen_missing_Diabetes_missing090


Dataset (missing=90%):   0%|          | 0/600 [00:00<?, ?it/s]

Map:   0%|          | 0/600 [00:00<?, ? examples/s]

/usr/local/lib/python3.12/dist-packages/peft/mapping_func.py:72: UserWarning: You are trying to modify a model with PEFT for a second time. If you want to reload the model with a different config, make sure to call `.unload()` before.
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


trainable params: 33,030,144 || all params: 4,055,498,240 || trainable%: 0.8145


Step,Training Loss
10,3.399821
20,1.514090
30,0.248506
40,0.107713
50,0.098739
60,0.097517
70,0.097117
80,0.094431
90,0.098466
100,0.092465


Обучение завершено за 5.8 мин
Найдено 20 checkpoints в /content/drive/MyDrive/finetuned_qwen_missing_Diabetes_missing090


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-19:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-19  ROC-AUC=0.5868


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-38:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-38  ROC-AUC=0.5631


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-57:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-57  ROC-AUC=0.5854


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-76:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-76  ROC-AUC=0.6229


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-95:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-95  ROC-AUC=0.6775


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-114:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-114  ROC-AUC=0.6844


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-133:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-133  ROC-AUC=0.6015


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-152:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-152  ROC-AUC=0.6345


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-171:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-171  ROC-AUC=0.6775


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-190:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-190  ROC-AUC=0.6139


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-209:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-209  ROC-AUC=0.6654


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-228:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-228  ROC-AUC=0.6506


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-247:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-247  ROC-AUC=0.6759


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-266:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-266  ROC-AUC=0.6327


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-285:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-285  ROC-AUC=0.6444


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-304:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-304  ROC-AUC=0.5718


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-323:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-323  ROC-AUC=0.6539


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-342:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-342  ROC-AUC=0.6177


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-361:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-361  ROC-AUC=0.6145


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-380:   0%|          | 0/154 [00:00<?, ?it/s]

  checkpoint-380  ROC-AUC=0.5512
Лучший checkpoint: /content/drive/MyDrive/finetuned_qwen_missing_Diabetes_missing090/checkpoint-114  (ROC-AUC=0.6844)


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Test eval:   0%|          | 0/154 [00:00<?, ?it/s]

Метрики:
  ROC-AUC: 0.6201
  F1: 0.3218
  Accuracy: 0.6169
  Precision: 0.4242
  Recall: 0.2593
Bootstrap (mean±std):
  ROC-AUC: 0.6183±0.0454
  F1: 0.3184±0.0644
  Accuracy: 0.6156±0.0392
  Precision: 0.4223±0.0883
  Recall: 0.2586±0.0584
Эксперимент завершен за 77.4 мин


In [ ]:
all_results

{'missing_000pct': {'metrics': {'ROC-AUC': np.float64(0.8038888888888889),
   'F1': 0.6333333333333333,
   'Accuracy': 0.7142857142857143,
   'Precision': 0.5757575757575758,
   'Recall': 0.7037037037037037},
  'bootstrap': {'ROC-AUC': '0.8037±0.0346',
   'F1': '0.6311±0.0524',
   'Accuracy': '0.7143±0.0364',
   'Precision': '0.5761±0.0621',
   'Recall': '0.7026±0.0621'},
  'time_total': 36.26706337928772,
  'time_per_sample': 0.23550041155381637,
  'best_checkpoint': '/content/drive/MyDrive/finetuned_qwen_missing_Diabetes_missing000/checkpoint-133',
  'eval_missing_rate': 0.0},
 'missing_020pct': {'metrics': {'ROC-AUC': np.float64(0.7559259259259259),
   'F1': 0.6382978723404256,
   'Accuracy': 0.6688311688311688,
   'Precision': 0.5172413793103449,
   'Recall': 0.8333333333333334},
  'bootstrap': {'ROC-AUC': '0.7549±0.0381',
   'F1': '0.6351±0.0487',
   'Accuracy': '0.6669±0.0386',
   'Precision': '0.5156±0.0550',
   'Recall': '0.8324±0.0508'},
  'time_total': 35.72437858581543,
  't

In [ ]:
print("\nBootstrap (mean±std):")
for tag, res in all_results.items():
    rate = tag.replace("missing_", "").replace("pct", "%").lstrip("0") or "0%"
    print(f"\n  missing={rate}")
    for k, v in res["bootstrap"].items():
        print(f"    {k}: {v}")


Bootstrap (mean±std):

  missing=%
    ROC-AUC: 0.8037±0.0346
    F1: 0.6311±0.0524
    Accuracy: 0.7143±0.0364
    Precision: 0.5761±0.0621
    Recall: 0.7026±0.0621

  missing=20%
    ROC-AUC: 0.7549±0.0381
    F1: 0.6351±0.0487
    Accuracy: 0.6669±0.0386
    Precision: 0.5156±0.0550
    Recall: 0.8324±0.0508

  missing=50%
    ROC-AUC: 0.7799±0.0357
    F1: 0.5846±0.0540
    Accuracy: 0.6886±0.0363
    Precision: 0.5488±0.0630
    Recall: 0.6302±0.0651

  missing=90%
    ROC-AUC: 0.6183±0.0454
    F1: 0.3184±0.0644
    Accuracy: 0.6156±0.0392
    Precision: 0.4223±0.0883
    Recall: 0.2586±0.0584


ROC-AUC (0.80): Хороший показатель для медицинской диагностики. Модель демонстрирует стабильную способность выявлять диабет. При 50% пропусков показатель удерживается на высоком уровне (0.78), что свидетельствует о сильной взаимосвязи между оставшимися клиническими признаками (например, уровнем глюкозы) и целевым состоянием.

Recall (0.70): Умеренная полнота при полных данных. Примечательно, что при 20% пропусков полнота возрастает до 0.83, что указывает на изменение порога уверенности модели в сторону более широкого охвата пациентов из группы риска. Однако при экстремальных 90% пропусков наблюдается резкий спад до 0.26, так как модель теряет критические биометрические показатели.

Precision (0.58): Точность предсказаний сбалансирована. Модель старается минимизировать ложноположительные диагнозы. При 20% пропусков точность снижается до 0.52, так как модель становится более «агрессивной» в поиске (что подтверждается ростом Recall), но при 90% пропусков она всё еще удерживает уровень 0.42.

Accuracy (0.71): Стабильная общая точность. Модель эффективно справляется с классификацией пациентов. Даже при 50% пропусков точность сохраняется на уровне 0.69, что подтверждает эффективность мультизадачного дообучения для работы с неполными медицинскими анкетами.

Время эксперимента: ~ 1 ч 10 мин.

Использовано оперативная памяти графического процесса: 50/80GB.

Графический процессор A100 80GB.